Implement changelog parsing and manipulation

nforro opened the pull request.

nforro commented:
I'm going to create some test cases as well.

FrNecas commented on specfile/changelog.py:
```diff
@@ -0,0 +1,127 @@
+        self.content = content.copy()
+
+    def __str__(self) -> str:
+        return f"{self.header}\n" + "\n".join(self.content) + "\n"
```
Couldn't this type of joining cause some problems with extra blanks lines? I am not completely sure whether they are allowed in the changelog or not. (same for `Changelog.reassemble()`)

nforro replied:
What extra blank lines do you mean? Could you give an example?

FrNecas replied:
What I thought about here was something like:
```
* Mon Jun 22 2022 Author theauthor@mail.com
<blank line here>
- entry
- entry
```

nforro replied:
Yes, this is actually valid, RPM doesn't have any issues with it. However, this would happen only if user wanted to have blank lines there, am I right? In that case, why not just let him?

FrNecas replied:
Oh, good point, for some reason I thought that the `__str__` method was being used for reassembly similar to sections. Looking at it again, the case mentioned above should hopefully work just fine as far as I can see (there will be a single changelog entry with `content` equal to `["", "- entry", "-entry"]`)

FrNecas commented on specfile/specfile.py:
```diff
@@ -104,3 +113,66 @@ def tags(self, section: str = "package") -> Iterator[Tags]:
+        Yields:
+            Spec file changelog as `Changelog` object or None if there is no %changelog section.
+        """
+        with self.sections() as sections:
```
Hm, this means that the changelog has to be parsed from scratch every time that the user of the library wants to change it, right? I had an idea for an alternative approach though I am not sure if it's really better. We could make `Changelog` inehrit from `Section` and it could have an `entries` attribute that would contain the list of entries similarly to the way it is now. This way, I guess it would be a bit easier to keep the two "views" on the changelog section (raw section and a list of changelog entries) in sync without having to parse it each time. What do you think about this approach?

FrNecas replied:
Makes sense, lets keep it the way as with `Tags`, thanks for the explanation

FrNecas commented on specfile/specfile.py:
```diff
@@ -104,3 +113,66 @@ def tags(self, section: str = "package") -> Iterator[Tags]:
+            changelog.insert(
+                0,
+                ChangelogEntry(
+                    date or datetime.date.today(), get_author(), evr, [f"- {entry}"]
```
This seems to assume that there is only a single "entry" in one changelog. But what if we wanted to create something like:
```
<header>
- something
- something else
```

nforro replied:
Thinking about it, I'd probably find more use cases for adding entries with mutiple lines myself. I'm going to change it.

lachmanfrantisek commented on README.md:
```diff
@@ -104,6 +104,16 @@ with specfile.tags() as tags:
+
+# working with changelog
+with specfile.changelog() as changelog:
+    # most recent changelog entry
+    print(changelog[0])
+    # making changes
+    changelog[1].date = datetime.date.today()
+    changelog[1].content.append('- another line')
+    # removing the oldest entry
+    del changelog[-1]
```
I am wondering if using a reverse order wouldn't be more consistent/stable.

e.g.:
```python
>>> len(the_changelog)
3
>>> the_changelog[2]
v1.0 changelog entry
>>> the_changelog.add_changelog_entry()
>>> the _changelog[2]
v2.0 changelog entry
```

The old-to-new ordering will return the same changelog entries in this example. And using `[-1]` can still be nicely used to get the last one.

nforro replied:
Makes sense, I'll change it.

lachmanfrantisek commented on specfile/changelog.py:
```diff
@@ -0,0 +1,127 @@
+
+from specfile.sections import Section
+
+DAYS = [
+    "Mon",
+    "Tue",
+    "Wed",
+    "Thu",
+    "Fri",
+    "Sat",
+    "Sun",
+]
+
+MONTHS = [
+    "Jan",
+    "Feb",
+    "Mar",
+    "Apr",
+    "May",
+    "Jun",
+    "Jul",
+    "Aug",
+    "Sep",
+    "Oct",
+    "Nov",
+    "Dec",
+]
```
Can't we use any library for this? I hope it's already defined somewhere.

nforro replied:
A library for this seemed like an overkill (I was thinking about [Babel](http://babel.pocoo.org/en/latest/), I'll try to find some lightweight alternative).

lachmanfrantisek commented on specfile/changelog.py:
```diff
@@ -0,0 +1,127 @@
+        """Entry header in the format that is being used in Fedora spec files."""
+        weekday = DAYS[self.date.weekday()]
+        month = MONTHS[self.date.month - 1]
+        return f"* {weekday} {month} {self.date:%d %Y} {self.author} - {self.evr}"
```
`rpmdev-bumpspec` uses time as well for me, e.g.:

```
 %changelog
* Mon Jan 24 14:13:29 CET 2022 Frantisek Lachman <flachman@redhat.com> - 0.5.3-1
- New upstream release 0.5.3
```

nforro replied:
You got me there, I didn't realize RPM supports this (since [4.14.0](https://github.com/rpm-software-management/rpm/commit/57f94a582602f0353cdb17a02dc12c4461d4f32d)). So we should definitely default to that.

lachmanfrantisek commented on specfile/changelog.py:
```diff
@@ -0,0 +1,127 @@
+        header_regex = re.compile(r"^\* \w+ (\w+) (\d{2}) (\d{4}) (.+) - (.+)$")
+        data = []
+        for line in section:
+            m = header_regex.match(line)
+            if m:
```
Can we use [walrus](https://www.python.org/dev/peps/pep-0572/) here? (What Python versions do we want to support?)

```suggestion
            if m := header_regex.match(line):
```

nforro replied:
I think we should support the same Python versions as packit. Are there plans to drop support for Python 3.6 and 3.7?

But nice idea, I didn't even think of it.

FrNecas replied:
Would be nice to use it but I recently used something specific from python 3.8 (where walrus operator was also added) and the CI was failing in EPEL.

lachmanfrantisek replied:
Good point with EPEL.

lachmanfrantisek reviewed (commented):
This looks really simple and clean, nice!

nforro commented:
I've addressed the comments, I decided to use [arrow](https://github.com/arrow-py/arrow) library for parsing and formatting timestamps.

I'm also going to write a test for `Specfile.add_changelog_entry()` (or should I do it later in another PR?).

lachmanfrantisek commented on specfile/changelog.py:
```diff
@@ -0,0 +1,128 @@
+                        legacy = True
+                else:
+                    legacy = False
+                data.insert(0, ChangelogEntry(datetime, author, evr, [], legacy))
```
Would it be possible to use keyword arguments to make this explicit? (I am speaking about the empty list.)
```suggestion
                data.insert(0, ChangelogEntry(datetime, author, evr, content=[], legacy=legacy))
```

lachmanfrantisek commented on specfile/specfile.py:
```diff
@@ -104,3 +115,72 @@ def tags(self, section: str = "package") -> Iterator[Tags]:
+            else:
+                return result
+
+        with self.changelog() as changelog:
```
:+1: for using a context manager --  both saving and manipulation looks really clear.

lachmanfrantisek commented on specfile/changelog.py:
```diff
@@ -0,0 +1,128 @@
+                data[0].content.append(line)
+        return Changelog(data)
+
+    def reassemble(self) -> List[str]:
```
I was a bit confused by the name.

What about `get_lines`/`get_raw`/`get_raw_sections`/`generate_text_repr`/`raw_content`/`raw_changelog`/`raw_section_lines` or anything else showing that we are exporting something and the result is in form of (raw) lines...

nforro replied:
I was trying to come up with a word that would be the opposite of "parse", but I'll replace it with something more clear.

lachmanfrantisek replied:
Thanks.

lachmanfrantisek reviewed (approved):
Looks really good to me!

Just two nit-picks...;)

FrNecas commented on specfile/changelog.py:
```diff
@@ -0,0 +1,130 @@
+            Constructed instance of `Changelog` class.
+        """
+        header_regex = re.compile(
+            r"^\* \w+ (\w+ \d{2} (?:\d{2}:\d{2}:\d{2} [A-Z]+ )?\d{4}) (.+) - (.+)$"
```
Nitpick: since this regex is getting a bit long (especially due to the date formats), it could be useful to use a verbose regex, would be a bit easier to understand. Or maybe at least add a comment with an example header line?

nforro replied:
I actually think using verbose regex would be less readable, I'll add example headers.

nforro replied:
It turns out there are other header formats in use, for example [kernel](https://src.fedoraproject.org/rpms/kernel/blob/rawhide/f/kernel.spec#_3005) has an unusual way to separate evr. I think we need to support this as well.

FrNecas reviewed (approved):
Looks clean to me, just one readability suggestion

FrNecas commented on specfile/changelog.py:
```diff
@@ -0,0 +1,134 @@
+
+class ChangelogEntry:
+    """
+     Class that represents a changelog entry.
```
There is an extra space in the docstring if I see correctly

FrNecas commented on specfile/changelog.py:
```diff
@@ -0,0 +1,134 @@
+    def extended_timestamp(self) -> bool:
+        """Whether the timestamp present in the entry header is extended (date and time)."""
+        try:
+            arrow.get(self.header, "ddd MMM DD YYYY")
```
Are you sure that this works correctly? Since you are passing the whole header to `arrow.get` now, this will always throw an exception, won't it? Since `self.header` is the whole `* <datetime> <author> <NVR>` string.

nforro replied:
No, the `get()` method actually searches for a match within the whole string (admittedly it's not made obvious in the official documentation, apart from several code examples), so it works as expected.

FrNecas commented on specfile/specfile.py:
```diff
@@ -104,3 +115,68 @@ def tags(self, section: str = "package") -> Iterator[Tags]:
+                    timestamp = now.date()
+            if author is None:
+                try:
+                    author = subprocess.check_output("rpmdev-packager").decode().strip()
```
This confused me a bit at first (setting email to None), maybe a comment mentioning that author from rpmdev-packager already contains the formatted mail could be useful.

nforro replied:
I'll do that. In fact, I'd prefer not separating e-mail and author's name and treat it as a single entity (RPM does so as well), but it was a requirement from packit.

FrNecas replied:
Good idea, I think that would be fine, the ChangelogEntry class will be simple and packit will have a high-level API for setting author and email separately. I was going to propose one more alternative approach (having an `Author` class that could be constructed either in the format of rpmdev-packager or from name and email) but that seems to be too complicated, especially from the caller's point of view so I'd go with what you proposed
